cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -72,20 +73,60 @@ public PipelinePlan plan(PipelineSpec spec) {
     // go through the stages and examine their plugin type to determine which stages are reduce stages
     Set<String> reduceNodes = new HashSet<>();
     Set<String> isolationNodes = new HashSet<>();
+    Set<String> actionNodes = new HashSet<>();
     Map<String, StageSpec> specs = new HashMap<>();
+
     for (StageSpec stage : spec.getStages()) {
       if (reduceTypes.contains(stage.getPlugin().getType())) {
         reduceNodes.add(stage.getName());
       }
       if (isolationTypes.contains(stage.getPlugin().getType())) {
         isolationNodes.add(stage.getName());
       }
+      if (Action.PLUGIN_TYPE.equals(stage.getPlugin().getType())) {
+        // Collect all Action nodes from spec
+        actionNodes.add(stage.getName());
+      }
       specs.put(stage.getName(), stage);
     }
 
+    // Map to hold set of stages to which there is a connection from a action stage.
+    Map<String, Set<String>> outgoingActionConnections = new HashMap<>();
```

Comment:
may be a little simpler to use a SetMultimap. Then you won't have to do the null checks.